components/authentication/authentication.js

```diff
@@ -28,31 +47,46 @@ async function postLogin(req,res){
         - use bcrypt to encode / decode it
         - if it's ok, send a token
     */
-
-    const user = await User.findOne({
-        email    : req.body.email,
-        password : req.body.password 
-    }).exec()
-
-    // user can be null.
-    if(!user){
-        throw "Either a user with that e-mail was not found or the password is wrong."
+    try{
+        const user = await User.findOne({
+            email    : req.body.email
+        }).exec()
+    
+        // user can be null.
+        if (!user)
+            throw new Error("User not found or password is incorrect.")
+
+        // compare hashedPassword with the provided password
+        let result = await bcrypt.compare(req.body.password, user.hashedPassword)
+
+        if(!result)
+            throw new Error("User not found or password is incorrect.")
+    
+        let { _id, firstName, lastName, email, role } = user
+    
+        // make a JWT
+        let token = jwt.sign(
+            {
+                id : _id,
+                firstName,
+                lastName,
+                email,
+                role
+            }, 
+            process.env.LOANAPP_JWT_SECRET,
+            {
+                expiresIn: process.env.LOANAPP_JWT_DURATION
+            }
+        )
+        // send it back
+        res.status(200).json({
+            tok : token
+        })

```

Comment:
I can see this as technical debt later. Are you comfortable abstracting the token logic to its own function or class?

Reply:
If there are plans on supporting other types of login (google, facebook, etc), maybe. It's a single function call. Why put more  things in the way?

Reply:
Just to conform to the separation of concerns principle. We could have a JWT class that handles creating tokens, validating them etc.

Reply:
I'll leave it there for now because I have no idea how to abstract it without making it harder to understand what is going on